packages/task/src/browser/quick-open-task.ts

```diff
@@ -186,36 +186,44 @@ export class QuickOpenTask implements QuickOpenModel, QuickOpenHandler {
         return `Task id: ${task.taskId}, label: ${task.config.label}`;
     }
 
-    private getFilteredTasks(recentTasks: TaskConfiguration[], configuredTasks: TaskConfiguration[], providedTasks: TaskConfiguration[])
-        : { filteredRecentTasks: TaskConfiguration[], filteredConfiguredTasks: TaskConfiguration[], filteredProvidedTasks: TaskConfiguration[] } {
+    private getFilteredTasks(recentTasks: TaskConfiguration[], configuredTasks: TaskConfiguration[], detectedTasks: TaskConfiguration[])
+        : { filteredRecentTasks: TaskConfiguration[], filteredConfiguredTasks: TaskConfiguration[], filteredDetectedTasks: TaskConfiguration[] } {
 
         const filteredRecentTasks: TaskConfiguration[] = [];
-        recentTasks.forEach(recent => {
-            const exist = [...configuredTasks, ...providedTasks].some(t => TaskConfiguration.equals(recent, t));
-            if (exist) {
-                filteredRecentTasks.push(recent);
+        for (const recent of recentTasks) {
+            const taskConfig = this.findConfig(recent.label, configuredTasks) || this.findConfig(recent.label, detectedTasks);
+            if (!taskConfig) {
+                continue;
             }
-        });
 
-        const filteredProvidedTasks: TaskConfiguration[] = [];
-        providedTasks.forEach(provided => {
-            const exist = [...filteredRecentTasks, ...configuredTasks].some(t => TaskConfiguration.equals(provided, t));
+            const exist = filteredRecentTasks.some(task => TaskConfiguration.equals(task, taskConfig));
             if (!exist) {
-                filteredProvidedTasks.push(provided);
+                filteredRecentTasks.push(taskConfig);
             }
-        });
+        }
 
         const filteredConfiguredTasks: TaskConfiguration[] = [];
         configuredTasks.forEach(configured => {
-            const exist = filteredRecentTasks.some(t => TaskConfiguration.equals(configured, t));
+            const exist = filteredRecentTasks.some(recent => TaskConfiguration.equals(configured, recent));
             if (!exist) {
                 filteredConfiguredTasks.push(configured);
             }
         });
 
-        return {
-            filteredRecentTasks, filteredConfiguredTasks, filteredProvidedTasks
-        };
+        const filteredDetectedTasks: TaskConfiguration[] = [];
+        detectedTasks.forEach(detected => {
+            const exist = filteredRecentTasks.some(recent => TaskConfiguration.equals(detected, recent)) ||
+                configuredTasks.some(configured => configured.label === detected.label);
+            if (!exist) {
+                filteredDetectedTasks.push(detected);
+            }
+        });
+
+        return { filteredRecentTasks, filteredConfiguredTasks, filteredDetectedTasks };
+    }
+
+    private findConfig(label: string, configs: TaskConfiguration[]): TaskConfiguration | undefined {
+        return configs.find(task => label === task.label);
```

Comment:
Not sure if the label comparison is enough. What if we have tasks 
- from different sources 
- and with the same label 
?

Reply:
All configurations from tasks.json file have the same source.
I tried to add a configuration with the same label to tasks.json file and got error:
 
![label_error](https://user-images.githubusercontent.com/5676062/59738627-94136b00-926a-11e9-864e-8987f4fe86e1.png)
Detected tasks from different sources can have the same label - you are right.  
 I'll try improve this behavior, thank you!

Reply:
fixed, please review

Reply:
i am testing right now